app/models/spree/product_set.rb

```diff
@@ -63,27 +63,37 @@ def update_product_variants(product, attributes)
 
   def update_product_master(product, attributes)
     return true unless attributes[:master_attributes]
-    update_variant(product, attributes[:master_attributes])
+    create_or_update_variant(product, attributes[:master_attributes])
   end
 
   def update_variants_attributes(product, variants_attributes)
     variants_attributes.each do |attributes|
-      update_variant(product, attributes)
+      create_or_update_variant(product, attributes)
     end
   end
 
-  def update_variant(product, variant_attributes)
+  def create_or_update_variant(product, variant_attributes)
     found_variant = product.variants_including_master.find do |variant|
       variant.id.to_s == variant_attributes[:id].to_s && variant.persisted?
     end
 
     if found_variant.present?
       found_variant.update_attributes(variant_attributes.except(:id))
     else
-      product.variants.create(variant_attributes)
+      create_variant(product, variant_attributes)
     end
   end
 
+  def create_variant(product, variant_attributes)
+      on_hand = variant_attributes.delete(:on_hand)
+      on_demand = variant_attributes.delete(:on_demand)
+
+      variant = product.variants.create(variant_attributes)
+
+      variant.on_demand = on_demand if on_demand.present?
+      variant.on_hand = on_hand.to_i if on_hand.present?
```

Comment:
I don't really understand why we assign them this way and can't be done through `#create`. Can you remind me why we need it?

Reply:
it's because these assignments will save the variant themselves and that messes up the overall variant save process. I think this is ok, first you have to create the variant and then you can manage stock (there are stock items and stock movements involved in stock management).

Reply:
uf, this is getting messy... we'll need to fix these things once get the upgrade live.

Reply:
I dont think it's too bad. you cant expect to create a variant and all it's parts in one mass_assignemnt. we should have a collection of services to do different things to models. I think it's totally fine to detach variant stock management from the variant model. the only thing we need to fix is to take these attributes out of the variant into a service. moving from variant_stock into some form of variant_stock_service.

so, in this case the actual change would be on the frontend where the "API" to create variants with stock settings disappears and the frontend needs to first create the variant and then set/manage stock levels (it would be a different API endpoint to manage stock levels).

Reply:
definitely, we'll need to work on the frontend and that will simplify the backend.